src/Algebra/Graph.hs

```diff
@@ -1066,3 +1105,79 @@ sparsify graph = res
         m <- get
         put (m + 1)
         overlay <$> s `x` m <*> m `y` t
+
+{- Note [The rules of foldg]
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+The rules for foldg work like this.
+
+* Up to (but not including) phase 1, we use the "buildG/f" rule to
+  rewrite all saturated applications of f with its buildG/foldg
+  form, hoping for fusion to happen (through the "foldg/buildG" rule).
+
+  In phase 1 and 0, we switch off these rules, inline buildG, and
+  switch on the "graph/f" rule, which rewrites the "foldg/f"
+  thing back into plain function if needed.
+
+  It's important that these two rules aren't both active at once
+  (along with build's unfolding) else we'd get an infinite loop
+  in the rules. Hence the activation control below.
+
+* R functions (R for Rewrite) are here to remember f after applying a
+  "builG/f" rule. These functions are higher-order functions and therefore
+  benefits from inlining.
+
+* The "fmapR/fmapR" rule optimises compositions of fmapR
+-}
+
+buildR :: forall a. F.Fold a -> Graph a
+buildR = F.foldg Empty Vertex Overlay Connect
+{-# INLINE [1] buildR #-}
+
+composeR :: (b -> c) -> (a -> b) -> a -> c
+composeR = (.)
+{-# INLINE [0] composeR #-}
+
+matchR :: b -> (a -> b) -> (a -> Bool) -> a -> b
+matchR e v p = \x -> if p x then v x else e
+{-# INLINE [0] matchR #-}
+
+flipR :: (b -> b -> b) -> (b -> b -> b)
+flipR = flip
+{-# INLINE [0] flipR #-}
+
+-- Rules to transform a buildR-equivalent function into its equivalent
+{-# RULES
+-- Transform a fmapR into its build equivalent
+"buildR/fmapR"  forall f g.
+  fmapR  f g = buildR (F.Fold $ \e v o c -> foldg e (composeR v f) o c g)
+
+-- Transform an induce into its build equivalent
+"buildR/induce"    [~1] forall p g.
+  induce p g = buildR (F.Fold $ \e v o c -> foldg e (matchR e v p) o c g)
+
+"buildR/foldg(fc)" [~1] forall (f::forall b. (b -> b -> b) -> (b -> b -> b))  g.
+  foldg Empty Vertex Overlay (f Connect) g = buildR (F.Fold $ \e v o c -> foldg e v o (f c) g)
+
+"buildR/foldg(fo)" [~1] forall (f::forall b. (b -> b -> b) -> (b -> b -> b))  g.
+  foldg Empty Vertex (f Overlay) Connect g = buildR (F.Fold $ \e v o c -> foldg e v (f o) c g)
+
+"buildR/foldg(fo)(hc)" [~1] forall (f::forall b. (b -> b -> b) -> (b -> b -> b)) (h::forall b. (b -> b -> b) -> (b -> b -> b)) g.
```

Comment:
I think we don't need this two-way rule, because it can be achieved by independent applications of the two rules above.

Reply:
I don't think so, imagine you have

```Haskell
foldg Empty Vertex (flipR Overlay) (flipR Connect)
```

None of the two single rules are matching with it...

Reply:
Oh, you are right. Not sure what I was thinking :)